fix(737): handle credentials resolution correctly

matthiasbruns opened the pull request:
On-behalf-of: SAP <matthias.bruns@sap.com>

<!-- markdownlint-disable MD041 -->
#### What this PR does / why we need it

During debugging issues with mounting .ocmconfig files into Docker containers, we had hard time finding the real root case for the issue.

In case we don’t find Docker a credential helper, but the config directs to a docker credential helper, we should fall back to a plain credential plugin / helper.


#### Which issue(s) this PR fixes
Contributes: https://github.com/open-component-model/ocm-project/issues/737

matthiasbruns commented on bindings/go/plugin/go.mod:
```diff
@@ -10,8 +10,8 @@ require (
 	ocm.software/open-component-model/bindings/go/constructor v0.0.0-20251024100521-a7be1bf5ce0d
-	ocm.software/open-component-model/bindings/go/credentials v0.0.2
-	ocm.software/open-component-model/bindings/go/descriptor/runtime v0.0.0-20251114112331-3af5ed79952f
+	ocm.software/open-component-model/bindings/go/credentials v0.0.3-0.20251114112331-3af5ed79952f
```
Still waiting for https://github.com/open-component-model/open-component-model/actions/runs/19364724866

fabianburth commented on bindings/go/repository/component/fallback/v1/repository.go:
```diff
@@ -328,19 +328,24 @@ func (f *FallbackRepository) GetResolvers() []*resolverruntime.Resolver {
+					slog.DebugContext(ctx, fmt.Sprintf("resolving credentials for repository %q failed: %s", specification, err.Error()))
+				} else {
+					return nil, fmt.Errorf("resolving credentials for repository %q failed: %w", specification, err)
+				}
```
Can we introduce a different API than programming against the error message string here?

matthiasbruns commented on cli/cmd/add/component-version/cmd.go:
```diff
@@ -444,7 +444,11 @@ func (prov *constructorProvider) GetTargetRepository(ctx context.Context, _ *con
+					slog.DebugContext(ctx, fmt.Sprintf("resolving credentials for repository %q failed: %s", prov.targetRepoSpec, err.Error()))
+				} else {
+					return nil, fmt.Errorf("resolving credentials for repository %q failed: %w", prov.targetRepoSpec, err)
+				}
```
https://github.com/open-component-model/open-component-model/pull/1191#discussion_r2537889881

`bindings/go/repository` does not know `bindings/go/credentials` and vice versa - I wanted to share the error types in both bit did not find a way to do so. The both depend on `runtime` - maybe we want a shared error package as well?

fabianburth commented on bindings/go/repository/component/fallback/v1/repository.go:
```diff
@@ -328,19 +328,24 @@ func (f *FallbackRepository) GetResolvers() []*resolverruntime.Resolver {
-			if credentials, err = f.credentialProvider.Resolve(ctx, consumerIdentity); err != nil {
-				slog.DebugContext(ctx, fmt.Sprintf("resolving credentials for repository %q failed: %s", specification, err.Error()))
+			if creds, err = f.credentialProvider.Resolve(ctx, consumerIdentity); err != nil {
+				// does not know the credentials package - so we need to check manually
```
Got it, thanks! Would importing it here lead to an import cycle? Or why don't you want to add it to the dependencies?

matthiasbruns commented:
# Current implementation

I added a dependency to repository in the credentials bindings
https://github.com/open-component-model/open-component-model/pull/1191/files#diff-03bd29972683701968a0cdc66dfc8c02f7e19c7fa685968bd8ca1fa67f0200e9R8

Which allows us to reuse the errors in
https://github.com/open-component-model/open-component-model/pull/1191/files#diff-94f4187b797a5b1ac093f9d76ca2ca05012ee3b3aa5e6e713510dff85d6cd8b3R120

## he downside

Credentials and repository are now coupled. Before, there was no dependency between those two.

# Better approach

Have a shared module, comparable to runtime, for shared models and errors, maybe also put it into runtime.
This is up for discussion, but I am not happy how it is done in ths pr right now.

Before I added a hard dependency, I did a string compare on the error in https://github.com/open-component-model/open-component-model/pull/1191/files#diff-247a834556eac3accf975570ae40723486445c03b9fec718adbf2e5807217a06R336

which is dirty but does not introduce coupling
